### ISSUE TITLE
Well-known protos are in src/google/protobuf/, but are imported from google/protobuf/

### ISSUE THREAD

**Issue by cgrushko** [google/protobuf#2598]
For example, `api.proto` contains

```
import "google/protobuf/source_context.proto";
import "google/protobuf/type.proto";
```

Bazel's `proto_library` expects to find these files relative to the `WORKSPACE` file, but they're in the `src/` subdirectory.

In practice, this makes it impossible for a user to import `google/protobuf/type.proto` and satisfy this dependency with `@com_google_protobuf//:well_known_protos` (where `com_google_protobuf` is the name of the remote repository the user gave the protobuf repo).

The reason is that `proto_library` will pass `-Isrc/google/protobuf/type.proto=src/google/protobuf/type.proto` to the proto-compiler.

**tl;dr**, is it possible to move the well-known protos one leve up in the repo?

**Comment by cgrushko** [comment]
@pherl for routing.

**Comment by pherl** [comment]
I guess we would need to use a genrule to copy the well known protos to one level up before passing them into the proto_library.

**Comment by cgrushko** [comment]
Out of curiosity, must they be in src/?

**Comment by pherl** [comment]
which language are you using? It seems we have python and cc proto_library.

**Comment by cgrushko** [comment]
This is about the language-agnostic `proto_library`.
If a `.proto` imports e.g., `google/protobuf/Any.proto`, that file must be in the `google/protobuf` directory relative to the workspace root.

**Comment by pherl** [comment]
I'm looking at the BUILD file in the repo. I don't think we have a language agnostic proto_library. Are you introducing your own?

**Comment by cgrushko** [comment]
Yes, correct. This will allow users of `java_proto_library` (and `java_lite` and `cc_`) to use `proto_library`'s that need to import the well-known protos.

**Comment by pherl** [comment]
Can do a genrule instead? Or we can change the filegroup to copy the files if that's easier. However, I'm not sure if there are any other projects using the filegroup already. It could be breaking for those projects.

**Comment by cgrushko** [comment]
Yes, I can use a `genrule` to move the files to the root, and there wrap them in a `proto_library`.
I'm curious if there's a reason not to move the `.proto` files up level up.

**Comment by cgrushko** [comment]
Ah, looks like the `genrule` idea won't work, because of https://github.com/bazelbuild/bazel/issues/2265.

**Comment by cgrushko** [comment]
@pherl can I send a PR to move the files?

**Comment by pherl** [comment]
I would avoid moving the files because:
1) the directory structure becomes confusing and less readable.
2) we have other build tools that rely on the path of those files. Moving them creates lots other issues.
3) there will be external projects depending on the path of the files as well.
4) it doesn't seem right to me to move the code just to make one build system happy.

That said, would you please wait for bazel to fix the issue? In the meanwhile, you can fork and copy the wkt's in your own repo.

**Comment by xfxyjwf** [comment]
@cgrushko 

I'm thinking we can probably introduce a proto_library attribute to do this. Something like:
```
proto_library(
  name = "well_known_types",
  srcs = ["src/google/protobuf/any.proto", ...],
  proto_paths = ["src"],
)
```
And then bazel will pass appropriate --proto_path flags to protocol compiler when compiling these protos or any dependent protos.

**Comment by cgrushko** [comment]
I think using a genrule is a considerably better option.
The less options the better, I think.

**Comment by xfxyjwf** [comment]
In the opensource world, it's very likely the directory layout of .proto file is not the same as the intended import path of these files. For example, many projects have a src/ directory just like protobuf. It would be more convenient if we don't have to write genrules to copy .proto files around.

**Comment by cgrushko** [comment]
If it's in src/ , why not import src/foo.proto?

**Comment by xfxyjwf** [comment]
Just like protobuf well known types, the import path is part of the public API and they have to be imported as "google/protobuf/any.proto" rather than "src/google/protobuf/any.proto". If they are imported using a different path, the resulted C++ code won't compile.

Protobuf has this problem and has to use genrule to copy the .proto files around. I think it's a fair assumption that other existing protobuf users who want to adopt bazel will get into the same problem. It's just nicer to have it built in bazel. But we can live with genrule for the time being though.

**Comment by haberman** [comment]
It doesn't sound like there is any action required here. I'm closing this bug, please re-open if further action is required.

**Comment by kamalmarhubi** [comment]
> In the opensource world, it's very likely the directory layout of .proto file is not the same as the intended import path of these files. For example, many projects have a src/ directory just like protobuf. It would be more convenient if we don't have to write genrules to copy .proto files around.

I agree with this. Particularly in a project built with Bazel where a common practice is to vendor libraries under `third_party`. In a non-Bazel vendored library, any proto libraries would need to have a path prefix removed.
